README.md
# pyhelpers

A toolkit for facilitating data manipulation

**Author**: Qian Fu [![Twitter URL](https://img.shields.io/twitter/url/https/twitter.com/Qian_Fu?label=Follow&style=social)](https://twitter.com/Qian_Fu)

[![PyPI](https://img.shields.io/pypi/v/pyhelpers?color=important&label=PyPI)](https://pypi.org/project/pyhelpers/)
[![PyPI - Python Version](https://img.shields.io/pypi/pyversions/pyhelpers?label=Python)](https://www.python.org/downloads/windows/)
[![GitHub](https://img.shields.io/github/license/mikeqfu/pyhelpers?color=green&label=License)](https://github.com/mikeqfu/pyhelpers/blob/master/LICENSE)
[![GitHub code size in bytes](https://img.shields.io/github/languages/code-size/mikeqfu/pyhelpers?color=yellowgreen&label=Code%20size)](https://github.com/mikeqfu/pyhelpers/tree/master/pyhelpers)
[![PyPI - Downloads](https://img.shields.io/pypi/dm/pyhelpers?color=yellow&label=Downloads)](https://pypistats.org/packages/pyhelpers)



---

**<span style="font-size:larger;">Contents</span>**

- [Installation](#installation)
- [Quick start - some examples](#quick-start)

---



## Installation <a name="installation"></a>

```
pip install pyhelpers
```

**Note:**

- Only a few frequently-used dependencies are installed through `pip install`. 
- When importing the module/functions whose dependencies are not available with the installation (or if you happen not to have those dependencies installed yet), an "*ModuleNotFoundError*" will be prompted and you may install them separately. 



## Quick start - some examples <a name="quick-start"></a>

The current version includes the following modules: 

- [`settings`](#settings)
- [`dir`](#dir)
- [`download`](#download)
- [`store`](#store)
- [`geom`](#geom)
- [`text`](#text)
- [`ops`](#ops)
- [`sql`](#sql)

Each of the modules includes a number of functions. For a quick start and demonstration purposes, only a couple of examples are provided below. 



## settings <a name="settings"></a>

This module can be used to change some common settings with [pandas](https://pandas.pydata.org/), [numpy](https://numpy.org/), [matplotlib](https://matplotlib.org/) and [gdal](https://pypi.org/project/GDAL/).


```python
from pyhelpers.settings import pd_preferences
```

The function `pd_preferences()` changes a few default [pandas](https://pandas.pydata.org/) settings (when `reset=False`), such as the display representation and maximum number of columns when displaying a `pandas.DataFrame`. 


```python
pd_preferences(reset=False)
```

If `reset=True`, all changed parameters are reset to their default values. Note that the preset parameters are for the module creator's preference; you can always change them in the source code to whatever suits your use. 



## dir <a name="dir"></a>

This module can be used to help manipulate directories.


```python
from pyhelpers.dir import cd
```

The function `cd()` returns the current working directory


```python
print(cd())
```

If you would like to direct to a customised folder, say `"test_dir"`, use `cd()` to change directory:


```python
path_to_folder = cd("test_dir", mkdir=False)
print(path_to_folder)
```

If the directory `path_to_folder` does not exist, setting `mkdir=True` (default: `False`), i.e. `cd("test_dir", mkdir=True)` will create it. 

More examples: (You could see the difference between `path_to_pickle` and `path_to_test_pickle` below)


```python
path_to_pickle = cd("test_dir", "dat.pickle")
print(path_to_pickle)
```


```python
path_to_test_pickle = cd("test_dir", "data", "dat.pickle")
# or, path_to_test_pickle == cd("test_dir\\data\\dat.pickle")
print(path_to_test_pickle)
```

Another function, `regulate_input_data_dir()`, may also be helpful sometimes:


```python
from pyhelpers.dir import regulate_input_data_dir

print(regulate_input_data_dir("test_dir"))
print(regulate_input_data_dir(path_to_test_pickle))
```



## download <a name="download"></a>

Note that this module requires [requests](https://2.python-requests.org/en/master/) and [tqdm](https://pypi.org/project/tqdm/). 


```python
from pyhelpers.download import download
```

Suppose you would like to download a Python logo from online where URL is as follows:


```python
url = 'https://www.python.org/static/community_logos/python-logo-master-v3-TM.png'
```

Firstly, specify where the .png file will be saved and what the filename is. For example, to name the downloaded file as `"python-logo.png"` and save it to a folder named `"images"`:


```python
python_logo_dir = cd("test_dir", "images")
path_to_python_logo = cd(python_logo_dir, "python-logo.png")
```

Then use `download()`


```python
download(url, path_to_python_logo)
```

You may view the downloaded picture by using [Pillow](https://pypi.org/project/Pillow/): 


```python
from PIL import Image

python_logo = Image.open(path_to_python_logo)
python_logo.show()
```

If you would like to remove the download directory, "../test_dir/images":


```python
from pyhelpers.dir import rm_dir
```

By setting `confirmation_required=True` (default), you will be asked to confirm whether you want to proceed if the folder is not empty:


```python
# Remove "picture" folder
rm_dir(cd("test_dir\\images"), confirmation_required=True)
```



## store <a name="store"></a>

This module can be used to help save and retrieve data. Note that some functions require [openpyxl](https://openpyxl.readthedocs.io/en/stable/), [XlsxWriter](https://xlsxwriter.readthedocs.io/) and [xlrd](https://xlrd.readthedocs.io/en/latest/).

Let's create a `pandas.DataFrame` first by using the above `xy_array`:


```python
import numpy as np
import pandas as pd

xy_array = np.array([(530034, 180381),   # London
                     (406689, 286822),   # Birmingham
                     (383819, 398052),   # Manchester
                     (582044, 152953)],  # Leeds
                    dtype=np.int64)

dat = pd.DataFrame(xy_array, columns=['Easting', 'Northing'])
```

If you would like to save `dat` as a [pickle](https://docs.python.org/3/library/pickle.html) file and retrieve it later, use `save_pickle` and `load_pickle`:


```python
from pyhelpers.store import save_pickle, load_pickle
```

To save `dat` to `path_to_test_pickle` (see [`dir`](#dir_py)):


```python
save_pickle(dat, path_to_test_pickle, verbose=True)  # default: verbose=False
```

To retrieve/load `dat` from `path_to_test_pickle`:


```python
dat_retrieved = load_pickle(path_to_test_pickle, verbose=True)
```

`dat_retrieved` and `dat`  should be identical:


```python
print(dat_retrieved.equals(dat))  # True
```

In addition, `store.py` also have functions for working with some other formats, such as

- `save_json()` and `load_json()` for .json
- `save_spreadsheet()` for .csv and .xlsx/.xls
- `save_multiple_spreadsheets()` and `load_multiple_spreadsheets()` for .xlsx/.xls
- `save_feather()` and `load_feather()` for .feather
- ...



## geom <a name="geom"></a>

This module can be used to help data manipulation related to geometric data and geographical data.

For example, if you need to convert coordinates from OSGB36 (British national grid) to WGS84 (latitude and longitude), use  `osgb36_to_wgs84`:


```python
from pyhelpers.geom import osgb36_to_wgs84
```

To convert a single coordinate, `xy`:


```python
xy = np.array((530034, 180381))  # London

easting, northing = xy
lonlat = osgb36_to_wgs84(easting, northing)  # osgb36_to_wgs84(xy[0], xy[1])

print(lonlat)  # (-0.12772400574286874, 51.50740692743041)
```

To convert an array of OSGB36 coordinates, `xy_array`:


```python
eastings, northings = xy_array.T
lonlat_array = np.array(osgb36_to_wgs84(eastings, northings))

print(lonlat_array.T) 
# [[-0.12772401 51.50740693]
#  [-1.90294064 52.47928436]
#  [-2.24527795 53.47894006]
#  [ 0.60693267 51.24669501]]
```

Similarly, if you would like to convert coordinates from latitude/longitude (WGS84) to easting/northing (OSGB36), import `wgs84_to_osgb36` instead.



## text <a name="text"></a>

This module can be used to help manipulate text data.

For example, suppose you have a `str` type variable, `string`:


```python
string = 'ang'
```

If you would like to find the most similar text to one of the following `lookup_list`:


```python
lookup_list = ['Anglia',
               'East Coast',
               'East Midlands',
               'North and East',
               'London North Western',
               'Scotland',
               'South East',
               'Wales',
               'Wessex',
               'Western']
```

Try the function `find_similar_str()`:


```python
from pyhelpers.text import find_similar_str
```

Setting `processor='fuzzywuzzy'` requires [fuzzywuzzy](https://github.com/seatgeek/fuzzywuzzy) (recommended) - `token_set_ratio`


```python
result_1 = find_similar_str(string, lookup_list, processor='fuzzywuzzy')

print(result_1)  # Anglia
```

Setting `processor='nltk'` requires [nltk](https://www.nltk.org/) - `edit_distance`


```python
result_2 = find_similar_str(string, lookup_list, processor='nltk', 
                            substitution_cost=1)

print(result_2)  # Anglia
```



## ops <a name="ops"></a>

This module gathers miscellaneous functions. 

For example, if you would like to request a confirmation before proceeding with some processes, use `confirmed`:


```python
from pyhelpers.ops import confirmed
```

You may specify, by setting `prompt`, the prompting message as to the confirmation:


```python
confirmed(prompt="Continue?...", confirmation_required=True)
# Continue?... [No]|Yes: # Type something
```

If you input `Yes` (or `Y`, `yes`, or something like `ye`), it should return `True`; otherwise, `False` (if the input being `No` or something like `n`). When `confirmation_required=False`, meaning that a confirmation is not required, this function would be null as it will just return `True`. 



## sql <a name="sql"></a>

This module provides a convenient way to establish a connection with a SQL server. The current version supports only [PostgreSQL](https://www.postgresql.org/) and a quick example is given below. 


```python
from pyhelpers.sql import PostgreSQL
```

#### (1) Connect to a database

Now you can connect a PostgreSQL server by specifying *host* (default: `'localhost'`), *port* (default: `5432`), *username* (default: `'postgres'`), *password* and *name of the database* (default: `'postgres'`)


```python
# Connect to 'postgres' (using all default parameters)
testdb = PostgreSQL(host='localhost', port=5432, username='postgres', 
                    password=None, database_name='postgres', verbose=True)
```

As `password=None`, you will be asked to input your password. And if `host`, `port`, `username` and `database_name` are all `None`, you will be asked to input manually as you run `PostgreSQL()`.

#### (2) Import and dump data into the database

After you have successfully established the connection, you could try to dump `dat` ([see above](#store)) into the database "postgres":


```python
testdb.dump_data(dat, table_name='test_table', schema_name='public', 
                 if_exists='replace', force_replace=False, 
                 chunk_size=None, col_type=None, method='multi', 
                 verbose=True)
```

The method `.dump_data()` relies on `pandas.DataFrame.to_sql()`; however, the default `method` is set to be `'multi'` (i.e. `method='multi'`) for a faster process. In addition,  `.dump_data()` further includes a callable [`psql_insert_copy`](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#io-sql-method), whereby the processing speed could be even faster.

For example:


```python
testdb.dump_data(dat, table_name='test_table', 
                 method=testdb.psql_insert_copy, verbose=True)
```

#### (3) Read data from the database

To retrieve the dumped data, use the method `.read_table()`:


```python
dat_retrieved = testdb.read_table('test_table')

print(dat.equals(dat_retrieved))  # True
```

Besides, there is an alternative way, which is more flexible with PostgreSQL statement (and could be faster especially when the table is fairly large): 


```python
sql_query = 'SELECT * FROM test_table'  # Or 'SELECT * FROM public.test_table'
dat_retrieved_ = testdb.read_sql_query(sql_query)
```

Note that `sql_query` should end without ';'


```python
print(dat.equals(dat_retrieved_))  # True
```

#### (4) Some other methods

To drop the table **'test_table'**, if `confirmation_required=True` (default: `False`), you will be asked to confirm whether you are sure to drop the table:


```python
testdb.drop_table('test_table', schema_name='public', 
                  confirmation_required=True, verbose=True)
# Confirmed to drop the table ... ? [No]|Yes: # Type something
# The table "test_table" has been dropped successfully.
```

If you would like to create your own database and name it as `'test_database'`:


```python
testdb.create_database("test_database", verbose=True)
# Creating a database "test_database" ... Done.
```

After successfully creating the database, you could check:


```python
test_db_exists = testdb.database_exists("test_database")
print(test_db_exists)  # True
```


```python
test_db_name = testdb.database_name
print(test_db_name)  # test_database
```

To drop this database, use the method `.drop_database()`. Setting `confirmation_required=True` (default and recommended) would require you to confirm whether you are sure to proceed:


```python
testdb.drop_database(confirmation_required=True, verbose=True)
# Confirmed to drop the database ... ? [No]|Yes: # Type something
# Dropping the database "test_database" ... Done.
```




RELEASES.md
### Release history

<br/>

#### [1.3.1](https://github.com/mikeqfu/pyhelpers/releases/tag/1.3.1)

(*10 February 2022*)

##### Main [changes](https://github.com/mikeqfu/pyhelpers/compare/1.3.0...1.3.1) since [1.3.0](https://pypi.org/project/pyhelpers/1.3.0/):

- Renamed the following functions: 
  - ~~`save()`~~ to `save_data()` in the module `store`;
  - ~~`get_user_agent_strings()`~~ to `load_user_agent_strings()` in the module `ops`.
- Improved the functions:
  - `remove_punctuation()`, `find_similar_str()` and `convert_md_to_rst()` in the module `text`;
  - `merge_dicts()` in the module `ops`.
- Added a few constants cached as the package/module is imported.
- Added the following new functions/methods: 
  - `load_csv()` and `load_data()` to the module `store`;
  - `numeral_english_to_arabic()` to the module `text`;
  - `PostgreSQL.list_schema_names()` to the module `dbms`;
  - `find_executable()` and `compare_dicts()` to the module `ops`.

**For more details, check out the [PyHelpers 1.3.1 documentation](https://pyhelpers.readthedocs.io/en/1.3.1/).**

<br/>

#### [1.3.0](https://github.com/mikeqfu/pyhelpers/releases/tag/1.3.0)

(*6 January 2022*)

##### Main [changes](https://github.com/mikeqfu/pyhelpers/compare/1.2.18...1.3.0) since [1.2.18](https://pypi.org/project/pyhelpers/1.2.18/):

- Renamed the following functions: 
  - in the module `dir`:
    - ~~`validate_input_data_dir()`~~ to `validate_dir()`;
  - in the module `ops`:
    - ~~`get_all_values_from_nested_dict()`~~ to `get_dict_values()`;
    - ~~`update_nested_dict()`~~ to `update_dict()`;
    - ~~`remove_multiple_keys_from_dict()`~~ to `remove_dict_keys()`.
- Improved the following functions/classes:
  - in the module `ops`:
    - `update_nested_dict()`;
    - `parse_size()`;
    - `download_file_from_url()`;
    - `init_requests_session()` (renamed from ~~`instantiate_requests_session()`~~);
  - in the module `text`:
    - `find_similar_str()`;
  - in the module `dbms`  (previously `sql`):
    - `PostgreSQL`;
  - in the module `store`:
    - `load_json()`;
  - in the module `geom`:
    - `wgs84_to_osgb36()`;
    - `osgb36_to_wgs84()`.
- Added the following functions to the module `ops`:
  - `parse_size()`;
  - `get_number_of_chunks()`;
  - `loop_in_pairs()`;
  - `is_url()`;
  - `update_dict_keys()`.
- Reduced the number of essential dependencies.

**For more details, check out the [PyHelpers 1.3.0 documentation](https://pyhelpers.readthedocs.io/en/1.3.0/).**

<br/>

#### [1.2.18](https://github.com/mikeqfu/pyhelpers/releases/tag/1.2.18)

(*20 October 2021*)

##### Main [changes](https://github.com/mikeqfu/pyhelpers/compare/1.2.17...1.2.18) since [1.2.17](https://pypi.org/project/pyhelpers/1.2.17/):

- In the module `ops`:
  - Fixed the issue in [`fake_requests_headers()`](https://github.com/mikeqfu/pyhelpers/commit/69adc0e7fdea6175c7ab6a9013470d541d96e3e2#diff-9266f224c0227114f50836ba656288a96bd7cca831301fe025c5a7cf9f4ab45aL1248-R1376) that may occasionally raise [`IndexError`](https://docs.python.org/3/library/exceptions.html#IndexError);
  - Added a new function [`get_fake_user_agent()`](https://github.com/mikeqfu/pyhelpers/commit/b4b822f71e03ea79ba4e663f2a20b772ce5824bd#diff-9266f224c0227114f50836ba656288a96bd7cca831301fe025c5a7cf9f4ab45aL1227-R1276).

**For more information about the update and/or a full walk-through of this release, <br/>check out [PyHelpers 1.2.18 documentation](https://pyhelpers.readthedocs.io/en/1.2.18/).**

<br/>

#### [1.2.17](https://github.com/mikeqfu/pyhelpers/releases/tag/1.2.17)

(*1 October 2021*)

##### Main [changes](https://github.com/mikeqfu/pyhelpers/compare/1.2.16...1.2.17) since [1.2.16](https://pypi.org/project/pyhelpers/1.2.16/):

- modified the function [fake_requests_headers](https://github.com/mikeqfu/pyhelpers/commit/39bf5c3bac09333f5ba043113dcbcc2706259ac6) to avoid raising [IndexError](https://docs.python.org/3/library/exceptions.html#IndexError)

**For more information about the update and/or a full walk-through of this release, <br/>check out [PyHelpers 1.2.17 documentation](https://pyhelpers.readthedocs.io/en/1.2.17/).**

<br/>

#### [1.2.16](https://github.com/mikeqfu/pyhelpers/releases/tag/1.2.16)

(*20 September 2021*)

##### Main [changes](https://github.com/mikeqfu/pyhelpers/compare/1.2.15...1.2.16) since [1.2.15](https://pypi.org/project/pyhelpers/1.2.15/):

- in the module [dir](https://github.com/mikeqfu/pyhelpers/blob/4cdf763ba7cecb46d66fbf3ea4eb684e04500d0e/pyhelpers/dir.py):
  - added a new function [go_from_altered_cwd](https://github.com/mikeqfu/pyhelpers/commit/a9e417669159bc7cc447cd966ee67120fffaf719#diff-9658546df62eded721ef6049c33f4b2d7e985d9cce8c08ba9538a32da5229a09R70-R117)
  - modified the function [cd](https://github.com/mikeqfu/pyhelpers/commit/a5be65408bca0a6aa279d7f785d22d27d2b6b277#diff-9658546df62eded721ef6049c33f4b2d7e985d9cce8c08ba9538a32da5229a09L15-R67)
  - renamed the function [~~is_dirname~~](https://github.com/mikeqfu/pyhelpers/commit/ebed45b34b276967ae34d893dd99f16cdce769d9#diff-9658546df62eded721ef6049c33f4b2d7e985d9cce8c08ba9538a32da5229a09L213) to [is_dir](https://github.com/mikeqfu/pyhelpers/commit/ebed45b34b276967ae34d893dd99f16cdce769d9#diff-9658546df62eded721ef6049c33f4b2d7e985d9cce8c08ba9538a32da5229a09R213)
- in the module [geom](https://github.com/mikeqfu/pyhelpers/blob/4cdf763ba7cecb46d66fbf3ea4eb684e04500d0e/pyhelpers/geom.py):
  - added a new function [drop_axis](https://github.com/mikeqfu/pyhelpers/commit/c141b6537f900e9f2ae595438af4a83e089fb4dd#diff-dc4a0e4af0eb0e5a833d868a38bde78a7c5736be4021d9ddf03b97c6f0cc8af8R426-R505)
  - modified the function [find_closest_points](https://github.com/mikeqfu/pyhelpers/commit/19423013ec565425f0c3dac8ffc50cceced14975#diff-dc4a0e4af0eb0e5a833d868a38bde78a7c5736be4021d9ddf03b97c6f0cc8af8L688-R805)
- in the module [ops](https://github.com/mikeqfu/pyhelpers/blob/4cdf763ba7cecb46d66fbf3ea4eb684e04500d0e/pyhelpers/ops.py):

pyhelpers/_cache.py
import importlib
import json

import pkg_resources


# An example DataFrame
def example_dataframe(osgb36=True):
    """
    Create an example dataframe.

    :param osgb36: whether to use data based on OSGB36 National Grid, defaults to ``True``
    :type osgb36: bool
    :return: an example dataframe
    :rtype: pandas.DataFrame
    """
    import pandas as pd

    if osgb36:
        _example_df = [
            (530034, 180381),  # London
            (406689, 286822),  # Birmingham
            (383819, 398052),  # Manchester
            (582044, 152953),  # Leeds
        ]
    else:
        _example_df = [
            (-0.12772401, 51.50740693),  # London
            (-1.90294064, 52.47928436),  # Birmingham
            (-2.24527795, 53.47894006),  # Manchester
            (0.60693267, 51.24669501),  # Leeds
        ]

    _index = ['London', 'Birmingham', 'Manchester', 'Leeds']
    _columns = ['Easting', 'Northing']

    _example_dataframe = pd.DataFrame(data=_example_df, index=_index, columns=_columns)

    return _example_dataframe


# import name: (package/module name, install name)
_OPTIONAL_DEPENDENCY = {
    "fuzzywuzzy": ("FuzzyWuzzy", "fuzzywuzzy"),
    "Levenshtein": ("python-Levenshtein", "python-Levenshtein"),
    "tqdm": ("tqdm", "tqdm"),
    "matplotlib": ("Matplotlib", "matplotlib"),
    "nltk": ("NLTK", "nltk"),
    "joblib": ("Joblib", "joblib"),
    "openpyxl": ("openpyxl", "openpyxl"),
    "xlsxwriter": ("XlsxWriter", "xlsxwriter"),
    "xlrd": ("xlrd", "xlrd"),
    "pyxlsb": ("pyxlsb", "pyxlsb"),
    "pyarrow": ("PyArrow", "pyarrow"),
    "orjson": ("orjson", "orjson"),
    "rapidjson": ("python-rapidjson", "python-rapidjson"),
    "osgeo.gdal": ("GDAL", "gdal"),
    "pyproj": ("pyproj", "pyproj"),
    "pypandoc": ("Pypandoc", "pypandoc"),
    "pdfkit": ("pdfkit", "pdfkit"),
}


def _import_optional_dependency(optional_dependency: str):
    # if optional_dependency not in sys.modules:
    try:
        importlib.import_module(optional_dependency)

    except (ImportError, ModuleNotFoundError):
        import_package_name = optional_dependency.split(".")[0]
        if import_package_name == 'osgeo':
            import_package_name = 'osgeo.gdal'

        package_name, install_name = _OPTIONAL_DEPENDENCY[import_package_name]
        msg = f"Missing optional dependency \"{package_name}\". " \
              f"Use pip or conda to install it, e.g. 'pip install {install_name}'."

        print(msg)


# == ops.py ======================================================================================

def _load_user_agent_strings():
    try:
        path_to_json = pkg_resources.resource_filename(__name__, "data/user-agent-strings.json")
        json_in = open(path_to_json, mode='r')
    except FileNotFoundError:
        json_in = open("pyhelpers/data/user-agent-strings.json", mode='r')

    user_agent_strings = json.loads(json_in.read())

    return user_agent_strings


_USER_AGENT_STRINGS = _load_user_agent_strings()


# == text.py =====================================================================================

def _english_written_numbers():
    metadata = dict()

    # Singles
    units = [
        'zero', 'one', 'two', 'three', 'four', 'five', 'six', 'seven', 'eight',
        'nine', 'ten', 'eleven', 'twelve', 'thirteen', 'fourteen', 'fifteen',
        'sixteen', 'seventeen', 'eighteen', 'nineteen', 'a',
    ]

    # Tens
    tens = ['', '', 'twenty', 'thirty', 'forty', 'fifty', 'sixty', 'seventy', 'eighty', 'ninety']

    # Larger scales
    scales = ['hundred', 'thousand', 'million', 'billion', 'trillion']

    # divisors
    metadata['and'] = (1, 0)

    # perform our loops and start the swap
    for i, word in enumerate(units):
        if word == 'a':
            metadata[word] = (1, 1)
        else:
            metadata[word] = (1, i)
    for i, word in enumerate(tens):
        metadata[word] = (1, i * 10)
    for i, word in enumerate(scales):
        metadata[word] = (10 ** (i * 3 or 2), 0)

    return metadata


_ENGLISH_WRITTEN_NUMBERS = _english_written_numbers()
